Cython/Compiler/Dataclass.py

```diff
@@ -0,0 +1,718 @@
+        while True:
+            name = "INIT_PLACEHOLDER_%s" % placeholder_count[0]
+            if (name not in placeholders
+                    and name not in fields):
+                # make sure name isn't already used and doesn't
+                # conflict with a variable name (which is unlikely but possible)
+                break
+            placeholder_count[0] += 1
+        return name
+
+    default_factory_placeholder = get_placeholder_name()
+    placeholders[default_factory_placeholder] = has_default_factory
+
+    function_body_code_lines = []
+
+    seen_default = False
+    for name, field in fields.items():
+        if not field.init.value:
+            continue
+        entry = node.scope.lookup(name)
+        if entry.annotation:
+            annotation = u": %s" % entry.annotation.string.value
+        else:
+            annotation = u""
+        assignment = u''
+        if field.default is not MISSING or field.default_factory is not MISSING:
+            seen_default = True
+            if field.default_factory is not MISSING:
+                ph_name = default_factory_placeholder
+            else:
+                ph_name = get_placeholder_name()
+                placeholders[ph_name] = field.default  # should be a node
+            assignment = u" = %s" % ph_name
+        elif seen_default:
+            error(entry.pos, ("non-default argument '%s' follows default argument "
+                              "in dataclass __init__") % name)
+            return "", {}, []
+
+        args.append(u"%s%s%s" % (name, annotation, assignment))
+
+        if field.is_initvar:
+            continue
+        elif field.default_factory is MISSING:
+            if field.init.value:
+                function_body_code_lines.append(u"    %s.%s = %s" % (selfname, name, name))
+        else:
+            ph_name = get_placeholder_name()
+            placeholders[ph_name] = field.default_factory
+            if field.init.value:
+                # close to:
+                # def __init__(self, name=_PLACEHOLDER_VALUE):
+                #     self.name = name_default_factory() if name is _PLACEHOLDER_VALUE else name
+                function_body_code_lines.append(u"    %s.%s = %s() if %s is %s else %s" % (
+                    selfname, name, ph_name, name, default_factory_placeholder, name))
+            else:
+                # still need to use the default factory to initialize
+                function_body_code_lines.append(u"    %s.%s = %s()"
+                                  % (selfname, name, ph_name))
+
+    args = u", ".join(args)
+    func_def = u"def __init__(%s):" % args
+
+    code_lines = [func_def] + (function_body_code_lines or ["pass"])
+
+    if node.scope.lookup("__post_init__"):
+        post_init_vars = ", ".join(name for name, field in fields.items()
+                                   if field.is_initvar)
+        code_lines.append("    %s.__post_init__(%s)" % (selfname, post_init_vars))
```

Comment:
This will need to handle keyword arguments right?

Reply:
The documentation says:

> It will normally be called as `self.__post_init__()`. However, if any `InitVar` fields are defined, they will also be passed to `__post_init__()` in the order they were defined in the class.

That matches what I do (I think). If you have a counter-example then let me know though!